samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -18,91 +18,74 @@
+  // TODO: MetadataCache timeout has to be 0 for the leader so that it can always have the latest information associated
+  // with locality. Since host-affinity is not yet implemented, this can be fixed as part of SAMZA-1197
+  private static final int METADATA_CACHE_TTL_MS = 5000;
 
   private final ZkUtils zkUtils;
   private final String processorId;
   private final ZkController zkController;
-  private final ScheduleAfterDebounceTime debounceTimer;
-  private final StreamMetadataCache  streamMetadataCache;
+
   private final Config config;
   private final CoordinationUtils coordinationUtils;
 
+  private StreamMetadataCache streamMetadataCache = null;
+  private ScheduleAfterDebounceTime debounceTimer = null;
   private JobCoordinatorListener coordinatorListener = null;
   private JobModel newJobModel;
 
   public ZkJobCoordinator(Config config) {
-    this.debounceTimer = new ScheduleAfterDebounceTime();
     this.config = config;
     this.processorId = createProcessorId(config);
     this.coordinationUtils = new ZkCoordinationServiceFactory()
         .getCoordinationService(new ApplicationConfig(config).getGlobalAppId(), String.valueOf(processorId), config);
     this.zkUtils = ((ZkCoordinationUtils) coordinationUtils).getZkUtils();
-    LeaderElector leaderElector = new ZkLeaderElector(this.processorId, zkUtils);
+    LeaderElector leaderElector = new ZkLeaderElector(processorId, zkUtils);
     leaderElector.setLeaderElectorListener(new LeaderElectorListenerImpl());
-
     this.zkController = new ZkControllerImpl(processorId, zkUtils, this, leaderElector);
-    streamMetadataCache = getStreamMetadataCache();
-  }
-
-  private StreamMetadataCache getStreamMetadataCache() {
-    // model generation - NEEDS TO BE REVIEWED
-    JavaSystemConfig systemConfig = new JavaSystemConfig(this.config);
-    Map<String, SystemAdmin> systemAdmins = new HashMap<>();
-    for (String systemName: systemConfig.getSystemNames()) {
-      String systemFactoryClassName = systemConfig.getSystemFactory(systemName);
-      if (systemFactoryClassName == null) {
-        String msg = String.format("A stream uses system %s, which is missing from the configuration.", systemName);
-        log.error(msg);
-        throw new SamzaException(msg);
-      }
-      SystemFactory systemFactory = Util.getObj(systemFactoryClassName);
-      systemAdmins.put(systemName, systemFactory.getAdmin(systemName, this.config));
-    }
-
-    return new StreamMetadataCache(Util.<String, SystemAdmin>javaMapAsScalaMap(systemAdmins), 5000, SystemClock.instance());
   }
 
   @Override
   public void start() {
+    streamMetadataCache = StreamMetadataCache.apply(METADATA_CACHE_TTL_MS, config);
+    debounceTimer = new ScheduleAfterDebounceTime(exception -> {
+        LOGGER.error("Received exception from in JobCoordinator Processing!");
+        stop();
+      });
+
     zkController.register();
   }
 
   @Override
-  public void stop() {
+  public synchronized void stop() {
```

Comment:
JobCoordinator can be stopped by StreamProcessor (can happen when the container thread stops/fails or when the user-api invokes a stop on streamprocessor) or the JobCoordinator's failure callback. So, it needs to be synchronized. 
Additionally, I think we need to maintain the state of the jobcoordinator or make stop idempotent. Atm, I didn't want to refactor JC state.